src/wbase/SendChannelShared.cc

```diff
@@ -298,69 +62,58 @@ bool SendChannelShared::buildAndTransmitResult(MYSQL_RES* mResult, int numFields
+    initTransmit(tMtxLock, *task);
 
-    numFields = mysql_num_fields(mResult);
     bool erred = false;
     size_t tSize = 0;
 
     int bytesTransmitted = 0;
     int rowsTransmitted = 0;
 
-    // If fillRows returns false, _transmitData is full and needs to be transmitted
+    // If fillRows returns false, transmitData is full and needs to be transmitted
     // fillRows returns true when there are no more rows in mResult to add.
     // tSize is set by fillRows.
     bool more = true;
     while (more && !cancelled) {
         util::Timer bufferFillT;
         bufferFillT.start();
-        more = !_transmitData->fillRows(mResult, numFields, tSize);
-        if (tSize > proto::ProtoHeaderWrap::PROTOBUFFER_HARD_LIMIT) {
-            LOGS_ERROR("Message single row too large to send using protobuffer");
-            erred = true;
-            util::Error worker_err(util::ErrorCode::INTERNAL,
-                                   "Message single row too large to send using protobuffer");
-            multiErr.push_back(worker_err);
-            break;
-        }
-        bytesTransmitted += _transmitData->getResultSize();
-        rowsTransmitted += _transmitData->getResultRowCount();
-        _transmitData->buildDataMsg(*task, largeResult, multiErr);
+        more = !transmitData->fillRows(mResult, tSize);
+        bytesTransmitted += transmitData->getResultSize();
+        rowsTransmitted += transmitData->getResultRowCount();
+        transmitData->buildDataMsg(*task, multiErr);
         bufferFillT.stop();
         bufferFillSecs += bufferFillT.getElapsed();
         LOGS(_log, LOG_LVL_TRACE,
              "buildAndTransmitResult() more=" << more << " " << task->getIdStr() << " seq=" << task->getTSeq()
-                                              << _dumpTr());
+                                              << dumpTransmit(tMtxLock));
 
         // This will become true only if this is the last task sending its last transmit.
-        // _prepTransmit will add the message to the queue and may try to transmit it now.
+        // prepTransmit will add the message to the queue and may try to transmit it now.
         bool lastIn = false;
         if (more) {
-            if (readRowsOk && !_prepTransmit(task, cancelled, lastIn)) {
+            if (!prepTransmit(tMtxLock, task, cancelled, lastIn)) {
                 LOGS(_log, LOG_LVL_ERROR, "Could not transmit intermediate results.");
-                readRowsOk = false;  // Empty the fillRows data and then return false.
                 erred = true;
                 break;
             }
         } else {
-            lastIn = transmitTaskLast(true);
+            lastIn = transmitTaskLast();
             // If 'lastIn', this is the last transmit and it needs to be added.
-            // Otherwise, just append the next query result rows to the existing _transmitData
+            // Otherwise, just append the next query result rows to the existing transmitData
```

Comment:
// Otherwise, just append the next query result rows to the existing transmitData
// and it will be sent later.

The comment should probably be changed to this.